### PR TITLE
feat(Core/Movement): Implement smooth waypoint transition

### PR DESCRIPTION
**Pull request by BarbzYHOOL**
* This is a draft, I put it here so it's not lost but I'm not gonna complete it myself nor claim the bounty, we're waiting for a dev to complete it :)
* Cherry Picked from TrinityCore https://github.com/TrinityCore/TrinityCore/pull/23864 made by @ovahlord

## CHANGES PROPOSED:
-  Mobs will turn smoothly (looks more realistic)

## ISSUES ADDRESSED:

There is a potential bounty on this https://github.com/azerothcore/forum/issues/61#issuecomment-575727273

## TESTS PERFORMED:
- Kito-p tested ingame but without mmaps and vmaps enabled. The result was that it worked, but some mobs began to act stupid and so it introduced a new bug (i have posted a gif on discord but actually discord is buggy so I can't link it)

## HOW TO TEST THE CHANGES:

- Follow any guard's path and he should turn smoothly

## Target branch(es):
- [x] Master

 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by Ovahlord** [comment]
rofl.

**Comment by pklloveyou** [comment]
I just tried it, and the creature was spinning in crazy circles. LOL

**Comment by BarbzYHOOL** [comment]
like this @pklloveyou https://imgur.com/a/8PiAcTr ? :rofl:

**Comment by pklloveyou** [comment]
I can't open that page，A closed Chinese network

File: src/server/game/Movement/MovementGenerators/WaypointMovementGenerator.cpp [hunk 1]
```diff
@@ -238,7 +259,7 @@ bool WaypointMovementGenerator<Creature>::DoUpdate(Creature* creature, uint32 di
         {
             // xinef: code to detect pre-empetively if we should start movement to next waypoint
             // xinef: do not start pre-empetive movement if current node has delay or we are ending waypoint movement
-            bool finished = creature->movespline->Finalized();
+            bool finished = creature->movespline->currentPathIdx() < _transitionPointId;
             if (!finished && !i_path->at(i_currentNode)->delay && ((i_currentNode != i_path->size() - 1) || repeating))
```
Comment (pklloveyou) [inline src/server/game/Movement/MovementGenerators/WaypointMovementGenerator.cpp]: - !finished
+ finished

Comment (pklloveyou) [reply src/server/game/Movement/MovementGenerators/WaypointMovementGenerator.cpp]: In doing so, the creature does not spin in circles

Comment (Ovahlord) [reply src/server/game/Movement/MovementGenerators/WaypointMovementGenerator.cpp]: have fun with your crashes by accessing splines without checking if they are actually in progress

Comment (pklloveyou) [reply src/server/game/Movement/MovementGenerators/WaypointMovementGenerator.cpp]: https://github.com/TrinityCore/TrinityCore/pull/23864 I got it from here

Comment (pklloveyou) [reply src/server/game/Movement/MovementGenerators/WaypointMovementGenerator.cpp]: image

Comment (Viste) [reply src/server/game/Movement/MovementGenerators/WaypointMovementGenerator.cpp]: and you dont see the code is different? just take and ok? lol

**Comment by kuramajssken001** [comment]
> I can't open that page，A closed Chinese network

你应该挂个VPN

File: src/server/game/Movement/MovementGenerators/WaypointMovementGenerator.cpp (hunk 1 above)
Comment (Ovahlord) [reply src/server/game/Movement/MovementGenerators/WaypointMovementGenerator.cpp]: Btw, you don't want that code. Really. It's far from optimal and before adding more hacky stuff, just let it be.

**Pull request closed without merging**